apps/meteor/client/components/AutoCompleteDepartment.tsx

```diff
@@ -0,0 +1,90 @@
+import { PaginatedSelectFiltered } from '@rocket.chat/fuselage';
+import { useDebouncedValue } from '@rocket.chat/fuselage-hooks';
+import { useTranslation } from '@rocket.chat/ui-contexts';
+import React, { memo, ReactElement, useMemo, useState } from 'react';
+
+import { useRecordList } from '../hooks/lists/useRecordList';
+import { AsyncStatePhase } from '../hooks/useAsyncState';
+import { useDepartmentsList } from './Omnichannel/hooks/useDepartmentsList';
+
+type AutoCompleteDepartmentProps = {
+	value: { value: string; label: string } | string;
+	onChange: (value: string) => void;
+	excludeDepartmentId?: string;
+	onlyMyDepartments?: boolean;
+	haveAll?: boolean;
+	haveNone?: boolean;
+};
+
+export const AutoCompleteDepartment = ({
+	value,
+	excludeDepartmentId,
+	onlyMyDepartments,
+	onChange,
+	haveAll,
+	haveNone,
+}: AutoCompleteDepartmentProps): ReactElement | null /* null added to satisfy register type condition */ => {
+	const t = useTranslation();
+	const [departmentsFilter, setDepartmentsFilter] = useState('');
+
+	const debouncedDepartmentsFilter = useDebouncedValue(departmentsFilter, 500);
+
+	const { itemsList: departmentsList, loadMoreItems: loadMoreDepartments } = useDepartmentsList(
+		useMemo(
+			() => ({
+				filter: debouncedDepartmentsFilter,
+				onlyMyDepartments,
+				haveAll,
+				haveNone,
+				excludeDepartmentId,
+			}),
+			[debouncedDepartmentsFilter, onlyMyDepartments, haveAll, haveNone, excludeDepartmentId],
+		),
+	);
+
+	const { phase: departmentsPhase, items: departmentsItems, itemCount: departmentsTotal } = useRecordList(departmentsList);
+
+	const findValue = useMemo<string>(() => (typeof value === 'string' ? value : value.value || ''), [value]);
+
+	const sortedByName = useMemo(
+		() =>
+			departmentsItems.sort((a, b) => {
+				if (a.value.value === 'all') {
+					return -1;
+				}
+
+				if (a.name > b.name) {
+					return 1;
+				}
+				if (a.name < b.name) {
+					return -1;
+				}
+
+				return 0;
+			}),
+		[departmentsItems],
+	);
+
+	const department = useMemo(() => sortedByName.find((dep) => dep._id === findValue)?.value, [sortedByName, findValue]);
```

Comment:
I think grouping the `findValue` variable with the sorting makes more sense here, since no other code needs it.
```
const department = useMemo(() => {
    const findValue = typeof value === 'string' ? value : value.value || '');
    return sortedByName.find((dep) => dep._id === findValue)?.value, [sortedByName, findValue]);
});
```

Reply:
ouch... in fact